akka-projection-cassandra/src/main/scala/akka/projection/cassandra/scaladsl/CassandraProjection.scala

```diff
@@ -64,6 +65,31 @@ object CassandraProjection {
       AtLeastOnce(1, Duration.Zero),
       GroupedHandlerStrategy(handler, groupAfterEnvelopes, groupAfterDuration))
 
+  /**
+   * Create a [[Projection]] with a [[Flow]] as the envelope handler. It has at-least-once processing semantics.
+   * The `Flow` should emit the same envelopes as it receives. The offsets in the emitted envelopes are stored
+   * in Cassandra. Since the offset is stored after processing the envelope it means that if the
+   * projection is restarted from previously stored offset some envelopes may be processed more than once.
+   *
+   * If the `Flow` filters out envelopes the corresponding offset will not be stored, and such envelope
+   * will be processed again if the projection is restarted and no later offset was stored.
+   *
+   * The `Flow` should not duplicate emitted envelopes (`mapConcat`) with same offset, because then it can result in
+   * that the first offset is stored and when the projection is restarted that offset is considered completed even
+   * though more of the duplicated enveloped were never processed.
+   */
```

Comment:
Do we foresee any other caveats with flow based?

I was looking at `FlowWithContext` as alternative but that can still filter and duplicate. Difference would be that the flow doesn't have to emit the Envelope since we could pass the Offset on the side, but I don't see it as problem for the user to pass on the Envelope since that is anyway the input.

Reply:
One aspect is that it would be wrong to use operators that re-order the elements, like `mapAsyncUnordered`. That could be a reason for limiting it to FlowWithContext. Wdyt @seglo ?

Reply:
If there is the possibility of breaking things, why are we offering this option? What are the advantages that, when weighted, surpass the disadvantages? 

If 1:1 is the only case that makes sense for projections, then that should be the only option.

Reply:
Yeah, I didn't think of reordering when looking at FlowWithContext first. It doesn't prevent filter and duplicates so then I didn't see the value. Let's go with FlowWithContext to prevent reordering.

Reply:
I think introducing the Flow API opens up a whole can of worms with message ordering, filtering, exploding, etc. We won't be able to guarantee 1:1 anymore and will have weird rules like in Lagom where you need to keep ratios the same (i.e. when filtering input you still have to emit the number of Done's from before the filter).

The Flow is obviously useful to power users, but until we can enforce constraints on streams (like some of @raboof 's ideas around context mapping) it might lead some users astray.

Reply:
But WithContext ensures that the order is kept, right?
Filtering isn’t a big problem unless all are filtered out? They will be consumed again, and filtered out again, but assuming that there are some passed through that shouldn’t grow too large. I agree that it must come with warnings.

Reply:
WithContext only ensures reordering by not including `mapAsyncUnordered` and other operators that reorder or do some kind of grouping.  The user can still drop down to regular flows if they want, which they will do because they're missing those operators in the first place.

Yes I guess as long as they emit some `Envelope` and don't filter all records then it will be good enough for at least once.

Reply:
I somewhat improved the documentation of `WithContext` at https://github.com/akka/akka/pull/29088/files to make these intentional restrictions more clear.

I think with the generalizations we discussed in https://github.com/akka/akka/pull/28712 `FlowWithContext` would fit very nicely for projections, but until we implement that its behavior might be a bit too much geared towards the Kafka use case to justify introducing it here.